### PR TITLE
API GW + API keys

### PR DESCRIPTION
**Pull request by prog893**
- Use APIGW instead of ALB for eden API
- Use usage plans + API key

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,15 +28,11 @@ module "eden" {
   name                  = "eden"
 
   # eden API ALB variables
```
Comment (toVersus) [inline README.md]:
```suggestion
   # eden API Gateway variables
```

File: README.md [hunk 2]
```diff
@@ -82,11 +86,11 @@ Optional query parameters:
 2019-04-08T20:32:06.310Z INFO     [main.py:create_alb_host_listener_rule:355] ELBv2 listener rule for target group arn:aws:elasticloadbalancing:ap-northeast-1:xxxxxxxxxxxx:targetgroup/dev-dynamic-test-create/b6918e6e5f10389d and host api-test-create.dev.example.com does not exist, will create new listener rule
```
Comment (toVersus) [inline README.md:86]: Can you replace sample logs with the current API Gateway version?

File: variables.tf [hunk 3]
```diff
@@ -49,41 +54,13 @@ variable "eden_table" {
...
-variable "api_access_logs_bucket_name" {
-  description = "S3 bucket name for saving eden API access logs"
-}
-
-variable "api_access_logs_prefix" {
-  description = "Path prefix for eden API access logs"
-}
-
 variable "api_acm_certificate_arn" {
   description = "ACM certificate ARN for eden API ALB"
```
Comment (toVersus) [inline variables.tf]:
```suggestion
   description = "ACM certificate ARN for eden API Gateway"
```

File: lambda/main.py [hunk 4]
```diff
@@ -26,139 +23,210 @@
...
+
+    table = dynamodb_resource.Table(table_name)
+
+    try:
+        r = table.query(
+            KeyConditionExpression=Key('type').eq('_profile') & Key('name').eq(profile_name)
+        )
+    except Exception as e:
+        if hasattr(e, 'response') and 'Error' in e.response:
+            logger.error(e.response['Error']['Message'])
```
Comment (toVersus) [inline lambda/main.py]:
```python
            error = e.response['Error']['Message']
            logger.error(error)
            return None, generate_response(error, 501)
```

File: lambda/main.py [hunk 5]
```diff
@@ -26,139 +23,210 @@
...
-    if 'name' not in request.args or 'image_uri' not in request.args:
-        return jsonify({
-            'status': 'error',
-            'reason': 'Insufficient query parameters',
-        }), 400
-
-    name: str = request.args.get('name')
-    image_uri: str = request.args.get('image_uri')
-    # noinspection PyBroadException
+            logger.error(f"Unknown exception raised: {e}")
```
Comment (toVersus) [inline lambda/main.py]:
```python
            error = f"Unknown exception raised: {e}"
            logger.error(error)
            return None, generate_response(error, 501)
```

File: lambda/main.py [hunk 6]
```diff
@@ -26,139 +23,210 @@
...
+    for p in check_params:
+        if p not in query_string_parameters or query_string_parameters[p] is None:
+            return generate_response(f"Necessary query string parameters not specified: {p}", 400)
+
+    return None
 
 
 def lambda_handler(event, context):
-    return serverless_wsgi.handle_request(app, event, context)
+    if event['queryStringParameters'] is not None:
```
Comment (toVersus) [inline lambda/main.py:183]: I prefer this one, but leave it to you:

```python
    profile_name = 'default'
    if event['queryStringParameters'] is not None:
        if 'profile' in event['queryStringParameters'] and event['queryStringParameters']['profile'] is not None:
            profile_name = event['queryStringParameters']['profile']
```

File: README.md (hunk 2 above)
Comment (prog893) [reply README.md:86]: Will do, but logs here should not change that much. We still are creating listener rules

File: lambda/main.py (hunk 6 above)
Comment (prog893) [reply lambda/main.py:183]: Yours looks better

**Review by toVersus — APPROVED** [review]
LGTM

**Pull request merged**